Multi PHP + Multi Environment + Check all commits

javiercasares opened the pull request:
This is the Cloudfest Hackathon 2024 PR.

This PR solves:

- Multi PHP. There is a new env option where you can set all the different PHP versions you want to test.
- Multi Environment. You can set a label to differentiate, for example, your “shared”, “vps”, or “cloud” setups.
- Check all commits. Until now, only the latest commit can be checked. With this new option, we can check the latest 10 commits so none of the commits are missing if they are approved without a lot of time between them.

The test-reporter will be upgraded to comply with these changes.

javiercasares committed:
environment for multi-php

javiercasares committed:
prepare for multi-php

javiercasares committed:
test for multi-php

javiercasares committed:
cleanup for multi-php

javiercasares committed:
report for multi-php

javiercasares committed:
temporary fix

javiercasares commented:
FYI: https://github.com/WordPress/phpunit-test-reporter/pull/109

javiercasares committed:
some changes

- add the label
- test dir is a dir (a tmp one)
- no cert validation by default
- simple WP by default

javiercasares committed:
env file with commints checking

javiercasares committed:
possible future files

javiercasares committed:
some changes

- support for labels
- read latest commits

javiercasares committed:
testing version

javiercasares committed:
fixes

javiercasares committed:
reorder the commits to run

javiercasares committed:
move the commit to the executed part

javiercasares commented:
The code is not very beauty, but it should work...

javiercasares committed:
checks that there are not duplicated tests to run

javiercasares committed:
Merge branch 'WordPress:master' into multi-php

javiercasares committed:
add some documentation

this is a first step to include some documentation inside the project, and not only in the handbook page.

pfefferle commented on docs/1-setup.md:
```diff
@@ -0,0 +1,333 @@
+
+## NodeJS installation
+
+If you are using Debian / Ubuntu, install or update NodeJS 18 with this command:
```
```suggestion
If you are using Debian / Ubuntu, install or update NodeJS 20 with this command:
```

pfefferle commented on docs/1-setup.md:
```diff
@@ -0,0 +1,333 @@
+# (Optionally) define the PHP executable to be called
+export WPT_PHP_EXECUTABLE=${WPT_PHP_EXECUTABLE-php}
+
+# (Optionally) array of versions (like: 8.0+/bin/php8.0,8.1+/bin/php8.1)
```
```suggestion
# (Optionally) array of versions (like: 8.0=/bin/php8.0;8.1=/bin/php8.1)
```

pfefferle commented on docs/1-setup.md:
```diff
@@ -0,0 +1,333 @@
+
+something like:
+
+`8.0+/bin/php8.0,8.1+/bin/php8.1`
```
```suggestion
`8.0=/bin/php8.0;8.1=/bin/php8.1`
```

pfefferle commented on docs/1-setup.md:
````diff
@@ -0,0 +1,333 @@
+Use as much versions as you want, but it will take more time. The idea is to put all the versions offered to users.
+
+```
+export WPT_PHP_EXECUTABLE_MULTI=7.4+/bin/php7.4,8.3+/bin/php8.3
````
```suggestion
export WPT_PHP_EXECUTABLE_MULTI=7.4=/bin/php7.4;8.3=/bin/php8.3
```

pfefferle commented on functions.php:
```diff
@@ -65,9 +65,11 @@ function perform_operations( $operations ) {
 	foreach( $operations as $operation ) {
 		log_message( $operation );
 		passthru( $operation, $return_code );
+		/* THIS IS A TEMPORARY (or not) COMMENTED CODE
```
???

javiercasares replied:
won't be there in next commits ;)

pfefferle commented on prepare.php:
```diff
@@ -50,6 +52,151 @@
+
+	foreach( $php_multi_versions as $php_multi_version ) {
+
+		$php_multi_v = explode( '+', $php_multi_version );
```
```suggestion
		$php_multi_v = explode( '=', $php_multi_version );
```

pfefferle commented on prepare.php:
```diff
@@ -50,6 +52,151 @@
+$WPT_PHP_EXECUTABLE_MULTI_ARRAY = array();
+if ( '' !== $WPT_PHP_EXECUTABLE_MULTI ) {
+
+	$php_multi_versions = explode( ',', $WPT_PHP_EXECUTABLE_MULTI );
```
```suggestion
	$php_multi_versions = explode( ';', $WPT_PHP_EXECUTABLE_MULTI );
```

pfefferle commented on report.php:
```diff
@@ -27,6 +27,7 @@
 $WPT_PREPARE_DIR    = trim( getenv( 'WPT_PREPARE_DIR' ) );
 $WPT_SSH_OPTIONS    = trim( getenv( 'WPT_SSH_OPTIONS' ) );
 $WPT_REPORT_API_KEY = trim( getenv( 'WPT_REPORT_API_KEY' ) );
+$WPT_PHP_EXECUTABLE_MULTI = trim( getenv( 'WPT_PHP_EXECUTABLE_MULTI' ) ) ? : '';
```
maybe align it by the `=`

pfefferle commented on prepare.php:
```diff
@@ -164,6 +276,7 @@
 	\$imagick_info = Imagick::queryFormats();
 }
 \$env = array(
+	'label'					 => '{$WPT_LABEL}',
```
there seems to be something broken with the indents?!

javiercasares committed:
Update .env.default

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
Update docs/1-setup.md

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
Update docs/1-setup.md

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
Update docs/1-setup.md

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
Update docs/1-setup.md

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
Update prepare.php

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
Update prepare.php

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

javiercasares committed:
TLS validation active by default

javiercasares committed:
Yoda and some organization and fixes

- Descriptions in 80 chars columns
- Some Yoda fixes
- Fix the "tab" vs "space"
- Check some "comments" that should not be there (they were because some errors on the developing era)

javiercasares committed:
Improving and normalizing code.

- Better comments / PHPDoc
- Improved code quality (do the same, in a better way)
- reduce duplicated code (that's something we didn't care at Cloudfest)
- updated the Runner Compatibity with PHP 7.2 (as WordPress is)

javiercasares committed:
80 columns

javiercasares committed:
readme changes

- Improved readability
- default .env updated from source
- other fixes

javiercasares committed:
documentation will be on the readme

javiercasares committed:
Documentation

- better documentation
- details on each value at the configuration file
- scripts so you can run (and update) the software automatically.

javiercasares commented:
I applied the suggestion @pfefferle did on WordCamp Europe, and reduce a lot of duplicated code that we couldn't do at the Cloudfest Hackathon, mostly on the parts when testing multiple PHP versions.

I did some improvements on the documentation, with some scripts so the software (from the main branch) is updated and maintained automatically.

I've been doing some tests in my local machine, and on my production server (where I was doing the former tests) and looks like everything is working.

On my side, I think everything is in order, and we can approve this and put it available for everyone.

benniledl commented:
### Running tests in parallel?
Running the PHPUnit test suite across 7 different PHP versions is taking a long time in my environment. Is there a way to run the tests in parallel to speed things up?

### Preparing a new environment for every version necessary?
Also, is it necessary to create a completely new environment for each PHP version? So far, I've been running the prepare.php file for the first PHP version, then report.php then using a script to search and replace the PHP version in the configuration files before rerunning the tests and reporting again. This approach seemed to work.

### Copying the environment + Running in parallel
I could see that using the same environment for multiple php versions in parallel could cause problems, is it maybe possible to just copy one environment to a different directory and then replace the php version in the config? Since preparing an environment also takes a long time for me.

WesRapyd commented:
What is the status of this pull ???

desrosj commented:
I'm going to put some time towards getting this tested and merged over the course of the next few weeks.

At Bluehost, we currently have a fork that we used to make changes directly and get our tests back up and running. However, it's a bit of a pain to maintain those small differences while keeping the fork in sync with changes here. Some of the changes actually unintentionally accomplish multi-php but just through the GitHub Action.

This PR is quite a bear, though. I am going to try and pull all of the unrelated documentation changes into a separate pull request, and all other unrelated functional changes into another in order to make this more focused.

desrosj reviewed (changes requested):
I've been extracting out some of the changes in this pull request that are unrelated to achieving multi PHP/multi environment/multi commit testing and tracking.

There are some merge conflicts that need to be resolved after (and if) #262, #256, and #255 are merged, but for now I've left some feedback on the parts specific to the goals of the PR.
